Add enabling of IPv6

joegsn opened the pull request:
I copied the description from the ZMQ description of ipv6 support in zmq_setsockopt.  Originally I thought to just copy the behavior for using integers 0/1, but it seemed reasonable to use false/true, given the boolean nature of the setting.  Thoughts?

paddor commented:
Cool! Here are my thoughts:
- test case is missing
- what's the default for the `ipv6` setting? (description)
- the reader method returns a boolean, and thus should end in `?`, e.g. `#ipv6?`

When adding the test case, you'll realize that the condition `new_value == true` is unnecessary / doesn't do what you meant. If you thought you wanna support Integers such as `0` and `1` as well, that condition doesn't do the trick. Take the `#router_mandatory=` method as an example:

``` ruby
      # Accept only routable messages on ROUTER sockets. Default is off.
      # @param bool [Boolean] whether to error if a message isn't routable
      #   (either if the that peer isn't connected or its SNDHWM is reached)
      def router_mandatory=(bool)
        Zsock.set_router_mandatory(@zocket, bool ? 1 : 0)
      end
```

I guess that's enough to process the input parameter.

I hope you agree with these points. :-)

paddor commented:
As for the test case, something like this in `spec/cztop/zsock_options_spec.rb` should do:

``` ruby
    describe "#ipv6=" do
      it "can enable IPv6" do
        expect(CZMQ::FFI::Zsock).to receive(:set_ipv6)
          .with(socket, 1)
        options.ipv6 = true
      end
      it "can disable " do
        expect(CZMQ::FFI::Zsock).to receive(:set_ipv6)
          .with(socket, 0)
        options.ipv6 = false
      end
    end
    describe "#ipv6?" do
      context "with default setting" do
        it "returns true" do
          assert_operator socket, :ipv6?
        end
      end
      context "when enabled" do
        it "returns true" do
          assert_operator socket, :ipv6?
        end
      end
      context "when disabled" do
        it "returns false" do
          refute_operator socket, :ipv6?
        end
      end
    end
```

Thank you.

joegsn commented:
I absolutely agree with your points.

I'm a bit unfamiliar with rspec.  Would you glance through this before I redo the pull request?

Basically, fixed it up for defaults, and added before-settings to the enabled/disabled entries.

``` ruby
    describe "#ipv6=" do
      it "can enable IPv6" do
        expect(CZMQ::FFI::Zsock).to receive(:set_ipv6)
          .with(socket, 1)
        options.ipv6 = true
      end
      it "can disable IPv6" do
        expect(CZMQ::FFI::Zsock).to receive(:set_ipv6)
          .with(socket, 0)
        options.ipv6 = false
      end
    end
    describe "#ipv6?" do
      context "with default setting" do
        it "returns false" do
          refute_operator options, :ipv6?
        end
      end
      context "with ipv6 enabled" do
        before(:each) {options.ipv6 = true}
        it "returns true" do
          assert_operator options, :ipv6?
        end
      end
      context "with ipv6 disabled" do
        before(:each) {options.ipv6 = false}
        it "returns false" do
          refute_operator options, :ipv6?
        end
      end
    end
```

paddor commented:
Looks good to me. You can just commit and push. The existing Pull Request will automatically be updated. :-)

paddor commented:
Thank you. :)

joegsn commented:
You're welcome.  Thanks for cztop.  I was going to give up on using zmq with Ruby before I learned about it.

paddor commented:
Wow, I really appreciate that. 👍
